### PR TITLE
Refactor eslint and prettier linter implementation

### PR DESCRIPTION
**Pull request by chrismaree**
I decided to spent a few hours looking into how to clean up our linter because it was really bugging me. I did not like these long string we had in our configs:
```
"prettier_vanilla": "$COMMAND --bracket-spacing 'core/**/*.js' 'common/**/*.js' '*.js' 'voter-dapp/src/**/*.js' 'sponsor-dapp-v2/src/**/*.js' 'core/contracts/**/*.sol' 'financial-templates-lib/*.js' 'liquidator/**/*.js' 'disputer/**/*.js'"
```
As well as all the nested, recursive calling. This PR changes all of that.

**Firstly**: why do we have _both_ eslint and prettier? Well they both have limitations on what they can lint and fix. Prettier is our main JS and solidity linter (with the help of `prettier-plugin-solidity`). eslint on the other hand is used to enforce other rules like spaces after a comment block (`//Coment...` BAD `// Comment...` GOOD) that prettier can't do on it's own. Eslint can, however, only run on js files for now. In an ideal world I'd like to just use one linter but right now we can't achieve all that we want with just one.

The following changes were made in this PR:
1. We now use a `pre-commit` hook over a `pre-push` hook. Originally I thought it would be less-invasive to have a pre-push hook. This is true in the happy case (code formatted correctly) but in the unhappy case this results in you needing to add an additional `lint` commit which is a pain. The way it is set up now will lint the files changed in your commit and add them to the stash that you plan to commit. This means that if you forgot to lint something and you then commit this will automagically deal with all linting and adding the changed files and you can continue without your workflow being interrupted.
2. husky hook uses both `lint-staged` and `pretty-quick` as they serve different purposes and look for different styling inconsistencies. However this is really fast to run! In my testing this hook takes less than 2 seconds to run. it also looks v cute in your console
image
3. Our `package.json` is _much_ cleaner now. I removed all the nested commands and replaced this with only _lint_ and _lint-fix_ which call the underlying `eslint` and `pretter` scripts.
4. We now use a `.prettierignore` and `.eslintignore` file over hard coding what to lint in the `package.json`. This is better because all code is now linted by default and only explicitly ignored files are left out.
5. There were a number of files that had gotten through the linter which this PR also fixes.

**TLDR:** The linter is now faster, more robust and easier to extend.

meme2

### CODE REVIEW

File: .eslintignore [hunk 1]
```diff
@@ -0,0 +1,8 @@
+node_modules
+coverage
+docs
+modules
+public
+ui
+sponsor-dapp-v2
```
Comment (chrismaree) [inline .eslintignore:7]: for now we cant run eslint in the dapp directories. this requires a number of react specific eslint plugins which I have not set up yet. Prettier still runs in these directories though.

File: .eslintrc.js [hunk 2]
```diff
@@ -2,23 +2,21 @@ module.exports = {
   env: {
     browser: true
   },
-  extends: "eslint:recommended",
+  extends: ["plugin:prettier/recommended"],
```
Comment (chrismaree) [inline .eslintrc.js:5]: This makes eslint use the prettier styling and config and as such we dont need to add all the ignore rules below like we had before.

File: .eslintrc.js [hunk 3]
```diff
@@ -2,23 +2,21 @@ module.exports = {
...
+  plugins: ["prettier"],
   parserOptions: {
-    ecmaVersion: 2018,
-    sourceType: "module"
+    ecmaFeatures: {
+      jsx: true,
+      modules: true
+    }
   },
+  parser: "babel-eslint",
```
Comment (chrismaree) [inline .eslintrc.js:13]: This was needed to enable us to parse certain es6 syntax we use in some of our off-chain infrastructure like the bots.

File: ci/lint.sh [hunk 4]
```diff
@@ -5,4 +5,4 @@ PROTOCOL_DIR=$(pwd)
 
 # Lint JS
 echo "Linting Solidity and js"
-npm run lint_check
+npm run lint
```
Comment (chrismaree) [inline ci/lint.sh:8]: I renamed this to make it cleaner. See the `package.json` file.

File: disputer/disputer.js [hunk 5]
```diff
@@ -86,7 +86,7 @@ class Disputer {
       } catch (error) {
         Logger.error({
           at: "Disputer",
-          message: `Failed to dispute liquidation`,
+          message: "Failed to dispute liquidation",
```
Comment (chrismaree) [inline disputer/disputer.js:89]: These code changes got through the linter before due to `--no-verify` pushes

File: package.json [hunk 6]
```diff
@@ -3,16 +3,10 @@
...
-    "lint_check": "npm run eslint_check && npm run prettier_check",
-    "lint_check_quick": "npm run eslint_check && npm run prettier_check_quick",
-    "eslint": "npm run eslint_vanilla -- --fix",
-    "eslint_check": "npm run eslint_vanilla",
-    "eslint_vanilla": "eslint 'core/**/*.js'",
-    "prettier": "COMMAND=prettier npm run prettier_vanilla -- --write",
-    "prettier_check": "COMMAND=prettier npm run prettier_vanilla -- --check",
-    "prettier_check_quick": "COMMAND=pretty-quick npm run prettier_vanilla -- --check",
-    "prettier_vanilla": "$COMMAND --bracket-spacing 'core/**/*.js' 'common/**/*.js' '*.js' 'voter-dapp/src/**/*.js' 'sponsor-dapp-v2/src/**/*.js' 'core/contracts/**/*.sol' 'financial-templates-lib/*.js' 'liquidator/**/*.js' 'disputer/**/*.js'"
+    "lint": "npm run eslint && npm run prettier -- --list-different",
```
Comment (chrismaree) [inline package.json:6]: This is much cleaner now. you can only `lint` or `lint-fix`. Both commands use eslint and prettier under the hood for their respective roles (see main PR comment).

File: package.json [hunk 7]
```diff
@@ -75,7 +73,10 @@
   },
   "husky": {
     "hooks": {
-      "pre-push": "npm run lint_check_quick"
+      "pre-commit": "echo '🕺  Running eslint on staged files' && lint-staged --allow-empty && echo '🏃‍♂️  Running pretty-quick on staged files' && pretty-quick --staged"
```
Comment (chrismaree) [inline package.json:76]: This is now a pre-commit hook. See PR comment for reasons.

File: package.json [hunk 8]
```diff
@@ -75,7 +73,10 @@
...
     }
+  },
+  "lint-staged": {
```
Comment (chrismaree) [inline package.json:79]: lint-staged lets us run eslint on staged files (pre-commit)

File: package.json (hunk 6 above)
Comment (mrice32) [reply package.json:6]: Just to triple check -- this command outputs a nonzero exit code if it finds issues, right?

Comment (chrismaree) [reply package.json:6]: Ye. this throws basically the same way that it did before. I updated the ci bash script as well.

File: .eslintrc.js (hunk 3 above)
Comment (nicholaspai) [reply .eslintrc.js:13]: Was this the reason that changes in the bots were throwing several errors even after we ran `npm run lint`?

File: package.json [hunk 9]
```diff
@@ -3,16 +3,10 @@
...
-    "eslint_check": "npm run eslint_vanilla",
-    "eslint_vanilla": "eslint 'core/**/*.js'",
-    "prettier": "COMMAND=prettier npm run prettier_vanilla -- --write",
-    "prettier_check": "COMMAND=prettier npm run prettier_vanilla -- --check",
-    "prettier_check_quick": "COMMAND=pretty-quick npm run prettier_vanilla -- --check",
-    "prettier_vanilla": "$COMMAND --bracket-spacing 'core/**/*.js' 'common/**/*.js' '*.js' 'voter-dapp/src/**/*.js' 'sponsor-dapp-v2/src/**/*.js' 'core/contracts/**/*.sol' 'financial-templates-lib/*.js' 'liquidator/**/*.js' 'disputer/**/*.js'"
+    "lint": "npm run eslint && npm run prettier -- --list-different",
+    "lint-fix": "npm run eslint -- --fix && npm run prettier -- --write",
+    "eslint": "eslint './**/*.js'",
+    "prettier": "prettier './**/*.js' './**/*.sol'"
```
Comment (mrice32) [inline package.json:9]: +1

File: disputer/disputer.js (hunk 5 above)
Comment (mrice32) [reply disputer/disputer.js:89]: hmmmm, but the linter should check *everything*...? husky should be a totally optional thing IIUC. Like we're not depending on it to catch errors that ci won't catch, right?

I'm assuming the explanation is that we didn't add this part of the repo to the prettier command (which was a mistake).

File: .eslintrc.js (hunk 2 above)
Comment (mrice32) [reply .eslintrc.js:5]: nice

File: disputer/disputer.js (hunk 5 above)
Comment (chrismaree) [reply disputer/disputer.js:89]: I thought so. I'm not sure how this go in but the linter in CI didnt catch it. You are right that husky is totally optional; if you choose to not use and and push anyways with incorrectly formatted code CI should throw it out from `lint.sh`. 

I think your explanation makes sense but it was specified in our `package.json` 🤔
